crates/metadata-store/src/grpc/client.rs

```diff
@@ -0,0 +1,253 @@
+use restate_core::network::net_util::CommonClientConnectionOptions;
+use restate_types::config::Configuration;
+use restate_types::net::AdvertisedAddress;
+use restate_types::nodes_config::NodesConfiguration;
+use restate_types::retries::RetryPolicy;
+use restate_types::storage::StorageCodec;
+use restate_types::Version;
+
+use crate::grpc::pb_conversions::ConversionError;
+use crate::grpc_svc::metadata_store_svc_client::MetadataStoreSvcClient;
+use crate::grpc_svc::{DeleteRequest, GetRequest, ProvisionRequest, PutRequest};
+
+/// Client end to interact with the metadata store.
+#[derive(Debug, Clone)]
+pub struct GrpcMetadataStoreClient {
+    channels: Arc<Vec<Channel>>,
+    svc_client: Arc<ArcSwap<MetadataStoreSvcClient<Channel>>>,
+}
+
+impl GrpcMetadataStoreClient {
+    pub fn new<T: CommonClientConnectionOptions>(
+        metadata_store_addresses: Vec<AdvertisedAddress>,
+        options: &T,
+    ) -> Self {
+        assert!(
+            !metadata_store_addresses.is_empty(),
+            "At least one metadata store needs to be configured"
+        );
+        let channels: Vec<_> = metadata_store_addresses
+            .into_iter()
+            .map(|address| create_tonic_channel_from_advertised_address(address, options))
+            .collect();
+        let svc_client = MetadataStoreSvcClient::new(
+            channels
+                .first()
+                .expect("at least one address mus be configured")
+                .clone(),
+        );
+
+        Self {
+            channels: Arc::new(channels),
+            svc_client: Arc::new(ArcSwap::from_pointee(svc_client)),
+        }
+    }
+
+    fn retry_policy() -> RetryPolicy {
+        Configuration::pinned()
+            .common
+            .network_error_retry_policy
+            .clone()
+    }
+
+    fn choose_different_endpoint(&self) {
+        // let's try another endpoint
+        let mut rng = rand::thread_rng();
+        let new_svc_client = MetadataStoreSvcClient::new(
+            self.channels
+                .choose(&mut rng)
+                .expect("at least one channel be present")
+                .clone(),
+        );
+        self.svc_client.store(Arc::new(new_svc_client))
+    }
+}
+
+#[async_trait]
+impl MetadataStore for GrpcMetadataStoreClient {
+    async fn get(&self, key: ByteString) -> Result<Option<VersionedValue>, ReadError> {
+        let retry_policy = Self::retry_policy();
+
+        let response = retry_on_network_error(retry_policy, || async {
+            let mut client = self.svc_client.load().as_ref().clone();
+
+            let response = client
+                .get(GetRequest {
+                    key: key.clone().into(),
+                })
+                .await
+                .map_err(map_status_to_read_error);
+
+            if response.as_ref().is_err_and(|err| err.is_network_error()) {
+                self.choose_different_endpoint();
+            }
```

Comment:
Requests that are rejected by non-leaders return an `Unavailable` status. That's why this + the retry policy will help us to find the actual leader by retrying. This also helps with dealing with metadata store nodes that are down. Very simplistic solution.